src/js/components/widgets/widget-dropdown.react.js

```diff
@@ -0,0 +1,118 @@
+     */
+    constructor (props) {
+        super(props);
+
+        this.bookmark = this.props.bookmark;
+        this.key = this.bookmark.widgetInfo.key;
+        this.value = '';
+
+        // If the dropdown is NOT of type source
+        if (this.key !== 'source') {
+            this.state = {
+                options: this.bookmark.widgetInfo.widgetMark.options
+            };
+        }
+        // If the dropdown is of type source
+        else {
+            // Try to find the sources from the tangram scene
+            let obj = getAddressSceneContent(tangramLayer.scene, this.bookmark.widgetInfo.widgetMark.source);
+            let keys = (obj) ? Object.keys(obj) : {};
+
+            // If the tangram scene has not yet loaded, set an empty options state in order for React to render
+            if (_.isEmpty(keys)) {
+                keys = [];
+            }
+
+            // Keys WILL NOT be empty in cases where users presses 'New' button on the same scene file.
+            // Keys WILL be empty when users reload the whole page
+            this.state = { options: keys };
+        }
+
+        this.handleChange = this.handleChange.bind(this);
+        this._setSource = this._setSource.bind(this);
+    }
+
+    /**
+     * React lifecycle function. Gets called once when DIV is mounted
+     */
+    componentDidMount () {
+        // Need to subscribe to when Tangram scene loads in order to populate the source widget
+        EventEmitter.subscribe('tangram:sceneinit', this._setSource);
+    }
+
+    /**
+     * Function called once the Tangram scene has loaded in order to update the source dropdown
+     */
+    _setSource () {
+        // If the dropdown is of type source then get sources from tangramLayer.scene
+        if (this.key === 'source') {
+            let obj = getAddressSceneContent(tangramLayer.scene, this.bookmark.widgetInfo.widgetMark.source);
+            let keys = (obj) ? Object.keys(obj) : {};
+
+            this.setState({ options: keys });
+        }
+    }
+
+    /**
+     * Called anytime there is a change in the dropdown form. I.e. when user opens or selects something
+     */
+    handleChange (e) {
+        this.value = e.target.value;
+
+        this._setEditorValue(this.value);
+    }
+
+    /* SHARED METHOD FOR ALL WIDGETS */
+    /**
+     *  Use this method within a widget to communicate a value
+     *  back to the Tangram Play editor.
+     */
+    _setEditorValue (string) {
+        this.bookmark = setCodeMirrorValue(this.bookmark, string);
+    }
+
+    /**
+     * Official React lifecycle method
+     * Called every time state or props are changed
+     */
+    render () {
+        return (
+            <FormGroup className='widget-dropdown' controlId='widget-form-dropdown'>
+                <FormControl componentClass='select' className='widget-form-control' placeholder='select' onChange={this.handleChange}>
+                    <option value='--select--'>-- select --</option>
```

Comment:
Can we set this `option` element to `disabled` to prevent it from being selected?